### ISSUE TITLE
OL-Cesium Layer Synchronization

### ISSUE THREAD

**Issue by schmidtk** [openlayers/ol3-cesium#139]
As a developer, it would be extremely useful to have more control over how ol3 layers are synchronized to Cesium. `olcs.VectorSynchronizer` and `olcs.RasterSynchronizer` make heavy use of closures to reuse a single synchronizer for every tile/vector layer, but this makes extending the class much more difficult.

I discussed two solutions to this with @ahocevar and we decided on the less invasive approach to the current framework that would give application developers significantly more freedom in customizing layer synchronization. Below are my thoughts on implementation.

`olcs.OLCesium` should have a root synchronizer implementing `olcs.ILayerSynchronizer` that can be overridden in the options to `olcs.OLCesium`. Replace the current `olcs.OLCesium` params with an options object containing the map, optional target, and optional layer synchronizer. The `olcs.ILayerSynchronizer` interface should have setters for the map/scene (or the `olcs.OLCesium`, since it has accessible references to both?) and a synchronize function, and the default synchronization class can create the existing `olcs.VectorSynchronizer` and `olcs.RasterSynchronizer`.

**Comment by ahocevar** [comment]
@schmidtk Considering your requirements and @gberaudo's comment, what about the following API:
```
var rasterSync = new olcs.RasterSynchronizer();
var vectorSync = new olcs.LayerSynchronizer({
  synchronize: function() {
    // custom synchronize function
  }
});
var ol3cesium = new olcs.OLCesium({
  map: map,
  synchronizers: [rasterSync, vectorSync]
});
```
My suggestion is that instead of creating an `olcs.ILayerSynchronizer`, we create an `olcs.LayerSynchronizer` which can either be overridden, or instantiated and configured with a `synchronize` function. The current `olcs.AbstractSynchronizer` would inherit from `olcs.LayerSynchronizer`.

**Comment by schmidtk** [comment]
Sorry for the slow response... I've been out for the past 2 weeks and playing catch up this morning.

I think providing a list of synchronizers in the `olcs.OLCesium` options would sufficiently solve this problem. It's not necessary to provide a custom synchronization function if the class can be overridden completely. At least for our purposes, we're interested in overriding `createSingleCounterpart` and the sychronization in `olcs.AbstractSynchronizer` can remain as-is.

This solution would require `olcs.OLCesium` to set the map/scene on each synchronizer prior to calling `synchronize` on each, but I don't see that being a problem.

**Comment by gberaudo** [comment]
> I think providing a list of synchronizers in the `olcs.OLCesium` options would sufficiently solve this problem. It's not necessary to provide a custom synchronization function if the class can be overridden completely. At least for our purposes, we're interested in overriding `createSingleCounterpart` and the sychronization in `olcs.AbstractSynchronizer` can remain as-is.
I like this solution, which avoid introducing another inheritance level.

> This solution would require `olcs.OLCesium` to set the map/scene on each synchronizer prior to calling `synchronize` on each, but I don't see that being a problem.
Instead, I would prefer an optional callback method which would create the synchronizer if defined.
That way, the application code is given the map and scene and is free to create its synchronizers
as it wants. We also avoid handling map and scene changes in the synchronizers.

The callback signature would look like `function create_synchronizers(map, scene): array_of_synchronizers`.

The olcs.OLCesium would be in charge of calling `synchronize` on each of the synchronizers.

Does it look good for you?

**Comment by ahocevar** [comment]
> We also avoid handling map and scene changes in the synchronizers.

If synchronizers have a non-API `setOLCesium` method which is only called by the OLCesium instance, we don't need to handle map and scene changes later on.

**Comment by schmidtk** [comment]
They could also have an `initialize(olCesium)` function in case they need to do some setup operations once they have a reference to the `olcs.OLCesium` object. Any of these sound fine to me.

**Comment by gberaudo** [comment]
I am not convinced by a method passing the OLCesium object to the synchronizers.
Especially in the OLCesium constructor, where it is not yet constructed.

**Comment by schmidtk** [comment]
Should they have a reference to the map/scene, or the OLCesium object? @ahocevar seemed to prefer the latter. And should the synchronizers be initialized in the constructor (as they are now), or when the OLCesium instance is enabled for the first time? Where do you propose your `create_synchronizers` method be called?

**Comment by ahocevar** [comment]
See #147 for a possible solution, using a callback function like @gberaudo suggested. I also tried to implement something along the lines of @schmidtk's most recent suggestion, but that would have required major changes in the way synchronizers are initialized.

@schmidtk, can you please see if #147 works for you?

**Comment by gberaudo** [comment]
The synchronizers should have a reference to the map and to the scene.
Of course, custom customizers may get other references too and even a
different constructor than the lib ones.

I was thinking of initializing the constructors at the same place as it
is now; which is also what @ahocevar implemented in #147.

It would be possible to instanciate them later but it would be more
complex. Do you think it would be useful?

**Comment by schmidtk** [comment]
@ahocevar That should work.
